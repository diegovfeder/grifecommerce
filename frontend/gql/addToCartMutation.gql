	mutation AddToCartMutation($id: ID!) {
		addToCart(productId: $id) {
			id
		}
	}
