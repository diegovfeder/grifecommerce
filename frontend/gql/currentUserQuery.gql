query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
