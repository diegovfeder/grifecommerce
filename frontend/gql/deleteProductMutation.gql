mutation DELETE_PRODUCT_MUTATION($id: ID!) {
	deleteProduct(id: $id) {
		id
		name
	}
}
