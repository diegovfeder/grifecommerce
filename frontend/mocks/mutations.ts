import DELETE_PRODUCT_MUTATION from '../gql/deleteProductMutation.gql';
import ADD_TO_CART_MUTATION from '../gql/addToCartMutation.gql';

const mockedMutations = {
	addToCart: {
		request: {
			query: ADD_TO_CART_MUTATION,
			variables: {
				id: '1',
			},
		},
		result: {
			data: {
				addToCart: {
					id: '1',
				},
			},
		},
	},
	deleteProduct: {
		request: {
			query: DELETE_PRODUCT_MUTATION,
			variables: {
				id: '1',
			},
		},
		result: {
			data: {
				deleteProduct: {
					id: '1',
					name: 'Sample Pack #01',
				},
			},
		},
	},
	signUp: {
		// TODO:
	},
};

export default mockedMutations;
