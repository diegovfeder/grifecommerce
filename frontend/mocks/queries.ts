import { CURRENT_USER_QUERY } from '../hooks/useUserQuery';

const mockQueries = {
	currentUser: {
		request: {
			query: CURRENT_USER_QUERY,
			variables: {},
		},
		result: {
			data: {
				authenticatedItem: {
					id: '1',
					email: 'diego@test.com',
					name: 'Diego Test',
					cart: {
						id: '1',
						quantity: '1',
						product: {
							id: '1',
							price: '200',
							name: 'Sample Pack #01',
							description: 'This is a sample pack',
							photo: {
								image: {
									publicUrlTransformed: 'https://via.placeholder.com/300x300',
								},
							},
						},
					},
				},
			},
		},
	},
};

export default mockQueries;
