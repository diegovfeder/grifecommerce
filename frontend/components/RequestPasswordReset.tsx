import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import StyledForm from './styles/StyledForm';
import useForm from '../hooks/useForm';
import Error from './ErrorMessage';
import {
	EventProps,
	RequestPasswordResetFormInputProps,
} from '../types/commonTypes';

const REQUEST_PASSWORD_RESET_MUTATION = gql`
	mutation REQUEST_PASSWORD_RESET_MUTATION($email: String!) {
		sendUserPasswordResetLink(email: $email) {
			code
			message
		}
	}
`;

const RequestPasswordReset = () => {
	const { inputs, handleChange, resetForm } =
		useForm<RequestPasswordResetFormInputProps>({
			email: '',
		});

	const [signup, { data, loading, error }] = useMutation(
		REQUEST_PASSWORD_RESET_MUTATION,
		{
			variables: inputs,
			// refetchQueries: [{ query: CURRENT_USER_QUERY }],
		},
	);

	const handleSubmit = async (e: EventProps) => {
		e.preventDefault(); // stop the form from submitting
		await signup().catch(console.error);
		resetForm();
	};

	return (
		<StyledForm method="POST" onSubmit={handleSubmit}>
			<h2>Request a Password Reset</h2>
			<Error error={error} />
			<fieldset>
				{data?.sendUserPasswordResetLink === null && (
					<p>Success! Check your email for a link!</p>
				)}

				<label htmlFor="email">
					Email
					<input
						type="email"
						name="email"
						placeholder="Your Email Address"
						autoComplete="email"
						value={inputs.email}
						onChange={handleChange}
					/>
				</label>
				<button type="submit">Request Reset!</button>
			</fieldset>
		</StyledForm>
	);
};

export default RequestPasswordReset;
