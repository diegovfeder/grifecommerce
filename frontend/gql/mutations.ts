import gql from 'graphql-tag';

export const SEND_USER_PASSWORD_RESET_LINK_MUTATION = gql`
	mutation SendUserPasswordResetLink($email: String!) {
		sendUserPasswordResetLink(email: $email)
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$email: String!
		$name: String!
		$password: String!
	) {
		createUser(data: { email: $email, name: $name, password: $password }) {
			id
			email
			name
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CreateOrder($token: String!) {
		checkout(token: $token) {
			id
			charge
			total
			items {
				id
				name
			}
		}
	}
`;
