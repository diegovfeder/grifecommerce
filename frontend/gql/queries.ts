import gql from 'graphql-tag';

export const USER_EMAIL_QUERY = gql`
	query UserEmail($email: String!) {
		user(where: { email: $email }) {
			id
			email
		}
	}
`;

export const PRODUCT_QUERY = gql`
	query Product($id: ID!) {
		product(where: { id: $id }) {
			id
			name
			price
			description
			photo {
				id
				altText
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const ALL_PRODUCTS_QUERY = gql`
	query AllProducts {
		products {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_QUERY = gql`
	query Products($take: Int, $skip: Int = 0) {
		products(take: $take, skip: $skip) {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_COUNT_QUERY = gql`
	query ProductsCount {
		productsCount
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
`;
