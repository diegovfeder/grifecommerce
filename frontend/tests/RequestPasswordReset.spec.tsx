import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';

import { resolveMockState } from './utils';
import { USER_EMAIL_QUERY } from '../gql/queries';
import { SEND_USER_PASSWORD_RESET_LINK_MUTATION } from '../gql/mutations';
import RequestPasswordReset from '../components/RequestPasswordReset';

const email = 'diego@test.com';
const mocks = [
	{
		request: {
			query: USER_EMAIL_QUERY,
			variables: { email },
		},
		result: {
			data: {
				user: {
					id: '1',
					email,
				},
			},
		},
	},
	{
		request: {
			query: SEND_USER_PASSWORD_RESET_LINK_MUTATION,
			variables: { email },
		},
		result: {
			data: {
				sendUserPasswordResetLink: true,
			},
		},
	},
];

describe('<RequestPasswordReset/>', () => {
	it('renders component and matches snapshot', () => {
		const { container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<RequestPasswordReset />
			</MockedProvider>,
		);
		expect(container).toMatchSnapshot();
	});

	it('calls the mutation when submitted', async () => {
		render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<RequestPasswordReset />
			</MockedProvider>,
		);
		userEvent.type(screen.getByPlaceholderText(/email/i), email);
		userEvent.click(screen.getByText(/Request Reset/));
		await resolveMockState();
		const success = await screen.findByText(/Password reset/i);
		expect(success).toBeInTheDocument();
	});
});
