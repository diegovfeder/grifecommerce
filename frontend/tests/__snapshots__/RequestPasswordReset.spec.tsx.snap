// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestPasswordReset/> renders component and matches snapshot 1`] = `
<div>
  <form
    autocomplete="new-password"
    class="StyledForm-sc-44wenk-0 iUkmmK"
    method="POST"
  >
    <h2>
      Request a Password Reset
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="new-password"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Reset!
      </button>
    </fieldset>
  </form>
</div>
`;
